ci/audit-logs-s3.py

```diff
@@ -1,126 +1,163 @@
+        "organization_name": get_cf_entity_name(
+            session, "organizations", audit_event["organization"]
+        ),
+        "space_name": get_cf_entity_name(session, "spaces", audit_event["space"]),
+    }
+
+
+# Upload a batch of audit events to S3 as a single object
+def upload_audit_events_to_s3(bucket_name, object_name, audit_events):
+    token = get_client_credentials_token()
+    with requests.Session() as s:
+        s.headers["Authorization"] = f"Bearer {token}"
+        body = "\n".join(
+            [
+                json.dumps(transform_audit_event_record(s, audit_event))
+                for audit_event in audit_events
+            ]
+        )
+    s3_client.put_object(
+        Bucket=bucket_name,
+        Key=object_name,
+        Body=body,
+        ContentType="application/json",
+    )
 
-def upload_to_s3(bucket_name, object_name, data):
-    try:
-        body = '\n'.join([
-                    json.dumps({
-                           **{k: v for k,v in item.items() if k not in ["links"]},
-                           "organization_name": get_cf_entity_name("organizations", f"{item['organization']}"),
-                           "space_name": get_cf_entity_name("spaces", f"{item['space']}")
-                           })
-                          for item in data
-                          ])
-        s3_client.put_object(
-            Bucket=bucket_name,
-            Key=object_name,
-            Body=body,
-            ContentType='application/json'
-            )
-        print("success for time "+ str(start_time))
-    except Exception as e:
-        print(traceback.format_exc())
-        print(f'Error upload file to S3 for time starting' + str(start_time) + " and end time" + str(end_time))
-        exit(1)
 
 def update_latest_stamp_in_s3(latest_timestamp):
     data = latest_timestamp
     s3_client.put_object(
-            Bucket=bucket_name,
-            Key=timestamp_key,
-            Body=data,
-            )
+        Bucket=bucket_name,
+        Key=timestamp_key,
+        Body=data,
+    )
+
+
+def get_start_end_time(now):
+    fifteen_minutes_ago = now - timedelta(minutes=15)
+    start_time = fifteen_minutes_ago.strftime("%Y-%m-%dT%H:%M:%SZ")
+    end_time = now.strftime("%Y-%m-%dT%H:%M:%SZ")
 
-def main():
     try:
-        audit_logs = get_audit_logs(start_time,end_time)
-        if audit_logs != "empty":
-            timestamp=audit_logs[-1]['created_at']
-            object_name = f"{now.year}/{now.month:02d}/{now.day:02d}/{now.hour:02d}/{now.minute:02d}/{now.second:02d}"
-            upload_to_s3(bucket_name, object_name,audit_logs)
-            update_latest_stamp_in_s3(timestamp)
+        current_stamp_response = s3_client.get_object(
+            Bucket=bucket_name, Key=timestamp_key
+        )
+        start_time = current_stamp_response["Body"].read().strip().decode("utf-8")
+        end_time = start_time + timedelta(minutes=15)
```

Comment:
What does this do?

Reply:
Previously, the logic was that the start time was either:

- fifteen minutes ago
- the time from the last saved timestamp in S3

But in both cases, we were using an end_time that was the current time of the script execution

This change makes it so that the end_time is 15 minutes after the start_time. But as I type that, I wonder if that would cause issues. So I'll remove it